include/envoy/stream_info/stream_debug_info.h

```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace StreamInfo {
+
+/**
+ * TODO(mergeconflict): document me
+ */
+class StreamDebugInfo {
```

Comment:
@mergeconflict FYI I was discussing this with @htuch and he seems to prefer just put this inside `StreamInfo`. I don't feel strongly about it. Can you sync up with him?

Reply:
Yeah, we chatted yesterday evening. I'm thinking to actually take the time to work on the matcher consolidation proposal (#5569) before coming back to this - I feel like that will clear this up.

Reply:
@mattklein123 Okay, I've convinced myself not to gate this on the matcher work. In this latest revision, I've removed `StreamDebugInfo` and replaced it with a simple struct, `Router::DebugConfig`, that has only the necessary configuration for the router-specific debugging features. I'm hoping this strikes the right balance between the flexibility of my initial metadata-only approach and the strong typing of your suggested `StreamDebugInfo` approach.

Let me know if this looks ok directionally, and I'll work on writing tests.

Reply:
I'm still not thrilled with using this map lookup code in a very hot path for a very rarely used feature, even though it is typed. Is there any way we can make this less computationally expensive in the case where there is no debug state? This is one of the reasons why I wanted this to be built-in.

/wait-any

Reply:
@mattklein123 since this is only intended for the router filter, this seems less of a candidate for a built-in, it's really using FilterState for the original intended purpose. That said, I agree it would make sense to make this fast. If there is no debug, we will miss on the map, so it's basically just a hash lookup and miss. Is the concern around additional memory accesses or cache use? There should be no allocations etc.

Reply:
Yes, valid point about the empty map being fast. I guess this is fine. I mainly do not like this pseudo-API, whether dynamic or static typed (marginally better). I will live with it for now I guess!

Reply:
Yeah, it's really a non-API. I will comment in the code that it's intended as temporary, until we make more progress on unifying matchers and debug configuration.